src/translator/html.cpp

```diff
@@ -240,11 +282,85 @@ size_t debugCountTokens(AnnotatedText const &text) {
   return tokens;
 }
 
+/// Helper function that consumes a tag as if it is a special tag, except that
+/// it takes nesting into account. I.e. `<a><a></a></a>` will be consumed to the
+// last `</a>`. Assumes TT_TAG_START is already consumed, which was necessary
+/// to determine whether this was an element that needed to be ignored.
+void consumeIgnoredTag(markup::Scanner &scanner, HTML::Tag &tag, std::string const &name) {
+  // Only full elements can be consumed this way. With void tags we don't know
+  // where to stop scanning. All other types cannot be nested anyway.
+  assert(tag.type == HTML::Tag::ELEMENT);
+
+  // TT_TAG_START is already consumed.
+  markup::Scanner::TokenType token;
+  size_t inside = 0;
+
+  // Consume the full open tag, i.e. all its attributes
+  while (!inside) {
+    token = scanner.next();
+    switch (token) {
+      case markup::Scanner::TT_ERROR:
+        ABORT("HTML parse error");
+      case markup::Scanner::TT_EOF:
+        ABORT("Did not find closing tag </{}>", name);
+      case markup::Scanner::TT_ATTRIBUTE:
+        tag.attributes += format(" {}=\"{}\"", scanner.attribute(), scanner.value());
+        break;
+      default:
+        // Not an attribute! Must be something inside the body or the closing
+        // tag already. Time to jump to the next loop.
+        ++inside;
+        break;
+    }
+  }
+
+  // Last token was something that would have triggered Scanner::scanBody(),
+  // which sets value() to start pointing at the body.
+  const char *start = scanner.start();
+
+  // Consume the rest of the HTML until (including) the final closing tag.
+  while (inside) {
+    switch (token) {
+      case markup::Scanner::TT_ERROR:
+        ABORT("HTML parse error");
+      case markup::Scanner::TT_EOF:
+        ABORT("Did not find closing tag </{}>");
+      case markup::Scanner::TT_TAG_START:
+      case markup::Scanner::TT_TAG_END:
+        // Note: Looking specifically for only our own type of tag so we don't
+        // have to care about whether other tags we encounter are void tags or
+        // not. Does assume the HTML is valid, as no stack is kept.
+        if (toLowerCase(scanner.tag()) == name) {
+          if (token == markup::Scanner::TT_TAG_END) {
+            if (--inside == 0) break;  // also stops loop because !inside
```

Comment:
`inside > 0` has to be true for the while-loop to start, so decrement should be safe. If decrement happens, it immediately jumps back to the `while (inside)` condition. I don't see a way for `--inside` to be called at that point.

Reply:
May I request a simplify for the cognitive load (This is probably my dislike for evil pointer arithmetic in conditions showing, and preference towards compiler optimizing the statements if there's room).

```cpp
if (token == markup::Scanner::TT_TAG_END) {
    --inside; // Action which will happen when we enter this block, not as part of the condition below.
    if (inside == 0) { // condition only evaluates what it should evaluate
    	break;  // also stops loop because !inside
    }
    ...
}
```

Reply:
I'll change it.

Note that `inside` is just a counter, no pointer magic. I wrote it as `if inside counts down to 0, stop` but I admit that it only reads that way once you know to read it that way.

Reply:
[`54be426` (#312)](https://github.com/browsermt/bergamot-translator/pull/312/commits/54be426b06943b6c05c283f977e05e2ca8c0f352) resolves this, looks like.